feat(3065): Ability to collapse and expand stage in the UI

sagar1312 opened the pull request:
## Context

Right now stage visualization default behavior is expanded.

<img width="1288" alt="image" src="https://github.com/user-attachments/assets/7952a396-c9f4-48b0-b465-42480d408eed" />


## Objective

Should have a way to collapse and expand a stage so it can minimize the real estate take in workflow graph.

**When Expanded**
<img width="1298" alt="image" src="https://github.com/user-attachments/assets/991ab811-b53e-4ab6-81d6-785b2b0729d1" />


**When Collapsed**
<img width="1293" alt="image" src="https://github.com/user-attachments/assets/d63b23f4-5943-4cc9-a8f0-7debb42a4988" />




## References

<!-- Links or resources that help clarify and support your intentions (e.g., Github issue) -->

## License

<!-- The following line must be included in your pull request -->

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

minghay reviewed (changes requested):
1. Doesn't work in V2 UI
2. Stages collapse to different heights
![Screenshot 2025-02-18 at 14-45-29 minghay_stages](https://github.com/user-attachments/assets/733a2a67-6679-4aa3-a02a-a0d4f71d034e)
3. Stages of the following form completely disappear
![Screenshot 2025-02-18 at 14-44-54 minghay_stages](https://github.com/user-attachments/assets/9084ca9f-a84c-4a76-b511-e3c1a53a06fa)
4. It's really difficult to tell what the collapse button indicates --- it looks a bit like a dot.  The black background also doesn't help so the whole styling around it should be revisited.
5. After collapsing, the stage options is completely broken and throws an error.

sagar1312 commented:
> 1. Doesn't work in V2 UI

Fixed

> 2. Stages collapse to different heights

It is as expected to acomodate the stage description.

> 3. Stages of the following form completely disappear

Fixed. Detached stages no longer disappears when collapsed.

> 4. It's really difficult to tell what the collapse button indicates --- it looks a bit like a dot.  The black background also doesn't help so the whole styling around it should be revisited.

Fixed the styling to make `-` more wider 

> 5. After collapsing, the stage options is completely broken and throws an error.

Fixed

@minghay Please take a look

minghay commented:
1. When a stage is collapsed, an error is still being thrown when selecting the pipeline options within the V2 UI

2. There's also another edge rendering issue with collapsed stages.  In the original uncollapsed state, you can see the edges are rendered without intersecting the job nodes (in particular the `pull` and `publish` jobs):
![Screenshot 2025-02-21 at 11-58-28 Events](https://github.com/user-attachments/assets/6d5e5bd5-178b-4321-a09c-392d82d18fbf)

When the stage is collapsed, the edges cut directly through the job node and actually merge into the previous edge somewhere after the arrowhead
![Screenshot 2025-02-21 at 11-59-00 Events](https://github.com/user-attachments/assets/1f7176d1-050c-4a92-93b0-ec1b0b9bb954)

minghay reviewed (changes requested):
Still a few outstanding issues:
1. The collapsed state of a stage is not preserved between graph updates in the V1 interface (i.e., build status changes update the graph and then stage that was previously collapsed is re-expanded during the update)
2. In the V1 interface, when a job node updates from running to complete and the following job is part of a collapsed stage, the graph will not update correctly)
3. In the V2 interface, the graph edges into and out of a stage are not rendered correctly
![Screenshot 2025-03-18 at 14-24-36 Events](https://github.com/user-attachments/assets/62052845-5129-43ac-a8b6-8c464268eb67)
![Screenshot 2025-03-18 at 14-40-07 Events](https://github.com/user-attachments/assets/71427280-9548-4719-b4c0-128acab0f6b6)
4. The stage border occasionally renders as as green before the stage has fully completed

sagar1312 commented:
> Still a few outstanding issues:
> 
> 1. The collapsed state of a stage is not preserved between graph updates in the V1 interface (i.e., build status changes update the graph and then stage that was previously collapsed is re-expanded during the update)

This is a new ask. The requirement to support this needs to hashed out. We should have a separate issue to track and implement this.
> 2. In the V1 interface, when a job node updates from running to complete and the following job is part of a collapsed stage, the graph will not update correctly)
> 3. In the V2 interface, the graph edges into and out of a stage are not rendered correctly
> 4. The stage border occasionally renders as as green before the stage has fully completed
Please share a sample pipeline to reproduce these

minghay commented:
> > 2. In the V1 interface, when a job node updates from running to complete and the following job is part of a collapsed stage, the graph will not update correctly)
> 
> Please share a sample pipeline to reproduce
> 
> > 4. The stage border occasionally renders as as green before the stage has fully completed
> 
> Please share a sample pipeline to reproduce

Seems like these are resolved now.

minghay reviewed (dismissed):
```
version: 4
shared:
  steps:
            - build: echo DONE
annotations:
    screwdriver.cd/chainPR: true

jobs:
    main:
        requires: [~commit]
    s:
        requires: []

    s2:
        requires: []
        
    d:
        requires: []
        
    d2:
        requires: []
        
    end:
        requires: [~stage@single, ~stage@with-setup-and-teardown, ~stage@detached, ~stage@detached-with-setup-and-teardown]

stages:
    single:
        jobs: [s]
        requires: [main]

    with-setup-and-teardown:
        jobs: [s2]
        requires: [main]
        setup:
            steps: 
                - build: echo 'More done'
        teardown:
            steps:
                - build: echo 'More done'
            
    detached:
        jobs: [d]
        
    detached-with-setup-and-teardown:
        jobs: [d2]
        setup:
            steps: 
                - build: echo 'More done'
        teardown:
            steps:
                - build: echo 'More done'
```

Collapsed stages may swap positions on the user.

In V2 UI:
1. If you collapse the `with-setup-and-teardown` stage, the `single` stage is unable to be collapsed.
2. If you collapse the `single` stage, then collapse the `with-setup-and-teardown` stage, the `with-setup-and-teardown` is unable to be re-expanded.
